lasagne/layers/recurrent.py

```diff
@@ -213,35 +230,88 @@ def __init__(self, incoming, input_to_hidden, hidden_to_hidden,
+                    hidden_to_hidden.output_shape[0]))
+
+        # Check that input_to_hidden and hidden_to_hidden output shapes match,
+        # but don't check a dimension of it's None for either shape
+        if not all(s1 is None or s2 is None or s1 == s2
+                   for s1, s2 in zip(input_to_hidden.output_shape[1:],
+                                     hidden_to_hidden.output_shape[1:])):
             raise ValueError("The output shape for input_to_hidden and "
-                             "input_to_hidden must be equal, but "
-                             "input_to_hidden.output_shape={} and "
-                             "hidden_to_hidden.output_shape={}".format(
+                             "hidden_to_hidden must be equal after the first "
+                             "dimension, but input_to_hidden.output_shape={} "
+                             "and hidden_to_hidden.output_shape={}".format(
                                  input_to_hidden.output_shape,
                                  hidden_to_hidden.output_shape))
 
+        # Check that input_to_hidden's output shape is the same as
+        # hidden_to_hidden's input shape but don't check a dimension of it's
+        # None for either shape
+        if not all(s1 is None or s2 is None or s1 == s2
+                   for s1, s2 in zip(input_to_hidden.output_shape[1:],
+                                     hidden_to_hidden.input_shape[1:])):
+            raise ValueError("The output shape for input_to_hidden "
+                             "must be equal to the input shape of "
+                             "hidden_to_hidden after the first dimension, but "
+                             "input_to_hidden.output_shape={} and "
+                             "hidden_to_hidden.input_shape={}".format(
+                                 input_to_hidden.output_shape,
+                                 hidden_to_hidden.input_shape))
+
         if nonlinearity is None:
             self.nonlinearity = nonlinearities.identity
         else:
             self.nonlinearity = nonlinearity
 
-        # Get the input dimensionality and number of units based on the
-        # expected output of the input-to-hidden layer
-        self.num_inputs = np.prod(input_shape[2:])
-        self.num_units = input_to_hidden.output_shape[-1]
+        # Get number of units based on the expected output of the
+        # hidden-to-hidden layer
+        self.hid_shape = input_to_hidden.output_shape
 
         # Initialize hidden state
         if isinstance(hid_init, T.TensorVariable):
-            if hid_init.ndim != 2:
+            if hid_init.ndim != len(self.hid_shape):
                 raise ValueError(
                     "When hid_init is provided as a TensorVariable, it should "
-                    "have 2 dimensions and have shape (num_batch, num_units)")
+                    "have the same shape as input_to_hidden.output_shape and "
+                    "hidden_to_hidden.output_shape")
```

Comment:
But `input_to_hidden.output_shape` and `hidden_to_hidden.output_shape` are not necessarily equal to begin with, depending on `precompute_input`. I'd just mention `hidden_to_hidden.output_shape` here and also set `self.hid_shape = hidden_to_hidden.output_shape` instead of `self.hid_shape = input_to_hidden.output_shape` (see https://github.com/Lasagne/Lasagne/pull/360#discussion_r36614207, that one seems to have disappeared... actually it seems you already updated the comment, but not the assignment).